src/Bedrock.Framework.Experimental/Protocols/Kafka/Messaging/KafkaRequestReaderWriter.cs

```diff
@@ -0,0 +1,29 @@
+using System;
+using System.Buffers;
+using Bedrock.Framework.Protocols.Kafka.Messaging.SchemaTypes;
+using Bedrock.Framework.Protocols.Kafka.Utils;
+
+namespace Bedrock.Framework.Protocols.Kafka.Messaging
+{
+    public class KafkaRequestReaderWriter : IMessageWriter<KafkaRequest>, IMessageReader<KafkaRequest>
+    {
+        public void WriteMessage(KafkaRequest message, IBufferWriter<byte> output)
+        {
+            var size = (int)message.RequestMessage.Length;
+
+            output.WriteInt(size);
+            output.WriteShort((short)message.ApiKey);
+            output.WriteShort(message.ApiVersion);
+            output.WriteInt(message.CorrelationId);
+            output.WriteNullableString(message.ClientId);
+
```

Comment:
Other than the nullable string you know the length here so it would be far more efficient to ask for a span that is (sizeof (int + short + short + int) and then write and slice the span with a single call to advance

Reply:
Forget me here your advance is on the buffer writter so it will just do a slice anyway and should be inlined...